doc/source/server_side_encryption.rst

```diff
@@ -0,0 +1,62 @@
+.. _server_side_encryption:
+
+########################
+Server-Side Encryption
+########################
+
+***********************
+Cloud
+***********************
+B2 cloud supports `Server-Side Encryption <https://www.backblaze.com/b2/docs/server_side_encryption.html>`_. All read
+and write operations provided by **b2sdk** accept encryption settings as an optional argument. Not supplying this
+argument means relying on bucket defaults - for **SSE-B2** and for no encryption. In case of **SSE-C**, providing an
+encryption key is crucial for successful downloading and copying.
+
+***
+API
+***
+Methods and classes that require encryption settings all accept an `EncryptionSetting` object, which holds information
+about present or desired encryption (mode, algorithm, key, key_id). Some, like copy operations, accept two sets of settings (for
+source and for destination). Sync, however, accepts an `EncryptionSettingsProvider` object, which is an
+`EncryptionSetting` factory, yielding them based on:
+
+* a :class:`b2sdk.v1.Bucket` object
+* file id
```

Comment:
that one is not yet known for files that are about to be uploaded. The `EncryptionSettingsProvider` interface is a mess and it needs to be fixed.

Reply:
Okay: here's my suggestion
```
@abstractmethod
    def get_setting_for_upload(
        self,
        bucket: Bucket,
        b2_file_name: str,
        file_info: Optional[dict],
        length: int,
    ) -> Optional[EncryptionSetting]:
        """
        Return an EncryptionSetting for uploading an object or None if server should decide.

        WARNING: the signature of this method is not final yet and not part of the public interface
        """

    @abstractmethod
    def get_source_setting_for_copy(
        self,
        bucket: Bucket,
        b2_file_id: str,
        b2_file_name: str,
        file_info: Optional[dict],
        length: int,
        mod_time_millis: int,
    ) -> Optional[EncryptionSetting]:
        """
        Return an EncryptionSetting for a source of copying an object or None if not required

        WARNING: the signature of this method is not final yet and not part of the public interface
        """

    @abstractmethod
    def get_destination_setting_for_copy(
        self,
        bucket: Bucket,
        b2_file_name: str,
        length: int,
    ) -> Optional[EncryptionSetting]:
        """
        Return an EncryptionSetting for a destination for copying an object or None if server should decide

        WARNING: the signature of this method is not final yet and not part of the public interface
        """

    @abstractmethod
    def get_setting_for_download(
        self,
        bucket: Bucket,
        b2_file_id: str,
        b2_file_name: str,
        file_info: Optional[dict],
        length: int,
        mod_time_millis: int,
    ) -> Optional[EncryptionSetting]:
        """
        Return an EncryptionSetting for downloading an object from, or None if not required

        WARNING: the signature of this method is not final yet and not part of the public interface
        """
```

Reply:
here `get_source_setting_for_copy` and for_download have a mandatory file name which I don't think is a good idea, the server doesn't actually require that. You can also not know what the bucket is!

I think many of those should be optional and None by default and Encryption provider should declare somehow what information it needs. If some layer (not sure what) will realize that this encryption provider doesn't get what it needs, it can try to retrieve it expensively on the fly and emit a warning, or it can fail with an error, if running in a more strict mode. By default UploadManager / CopyManager / Sync, or EncryptionProviderHelper (?) may refuse to call b2_get_file_info when the EncryptionProvider requires some piece of information they don't have when they are calling it, but a future implementation of EncryptionProviderHelper may also allow to do as much b2_get_file_info as needed or even figure out the pattern behind the subsequent calls and amortize the cost of retrieving the necessary information.

We need to realize that UploadManager, CopyManager and Sync are the only implementations that are calling encryption providers, but it's not going to stay like this forever. This class can be used in all sorts of ways by fuse drivers and whatnot.